File: easyapi_java/src/main/java/com/hlab/easyapi_java/easyapi/main/api/EasyApiCall.java
package com.hlab.easyapi_java.easyapi.main.api;

import android.support.annotation.NonNull;
import com.hlab.easyapi_java.easyapi.util.ResponseHandler;
import io.reactivex.Observable;
import io.reactivex.functions.Consumer;
import io.reactivex.observers.DisposableObserver;
import retrofit2.Call;

public interface EasyApiCall<T> {
  default EasyApi<T> initCall(@NonNull Call<T> call) {
    return null;
  }

  default EasyApiRx<T> initCall(@NonNull Observable<T> call) {
    return null;
  }

  default void executeWith(@NonNull DisposableObserver<T> observer) {
  }

  default void executeWith(@NonNull Consumer<T> successConsumer,
      @NonNull Consumer<Throwable> failureConsumer) {
  }

  void execute(boolean showErrorMessages, @NonNull ResponseHandler<T> responseHandler);

  void dispose();
}

File: easyapi_kotlin/src/main/java/com/hlab/easyapi_kotlin/easyapi/main/api/EasyApiCall.kt
package com.hlab.easyapi_kotlin.easyapi.main.api

import io.reactivex.Observable
import io.reactivex.functions.Consumer
import io.reactivex.observers.DisposableObserver
import retrofit2.Call

interface EasyApiCall<T> {
  fun initCall(call: Call<T>): EasyApi<T>? {
    return null
  }

  fun initCall(call: Observable<T>): EasyApiRx<T>? {
    return null
  }

  fun executeWith(observer: DisposableObserver<T>) {}

  fun executeWith(
    successConsumer: Consumer<T>,
    failureConsumer: Consumer<Throwable>
  ) {
  }

  fun execute(
    showErrorMessages: Boolean,
    responseHandler: (response: T?, isError: Boolean) -> Unit/*responseHandler: ResponseHandler<T>*/
  )

  fun dispose()
}
